File: assignment_4/kotlin/main.kt
fun main() {
    println("[Adapter Pattern]")
    val adapter = Adapter(ExternalClass())
    println(adapter.fetch())

    println()
    println("[Decorator Pattern]")
    val basicCoffee = BasicCoffee()
    println("basicCoffee = BasicCoffee()")
    println("basicCoffee.cost() = ${basicCoffee.cost()}") // 5

    val coffeeWithMilk = MilkDecorator(basicCoffee)
    println("\ncoffeeWithMilk = MilkDecorator(basicCoffee)")
    println("coffeeWithMilk.cost() = ${coffeeWithMilk.cost()}") // 7

    val coffeeWithSugar = SugarDecorator(basicCoffee)
    println("\ncoffeeWithSugar = SugarDecorator(basicCoffee)")
    println("coffeeWithSugar.cost() = ${coffeeWithSugar.cost()}") // 6

    val coffeeWithMilkAndSugar = SugarDecorator(coffeeWithMilk)
    println("\ncoffeeWithMilkAndSugar = SugarDecorator(coffeeWithMilk)")
    println("coffeeWithMilkAndSugar.cost() = ${coffeeWithMilkAndSugar.cost()}") // 8

    println()
    println("[Facade Pattern]")
    val myPC = Computer()
    myPC.boot()

    println()
    println("[Factory Method Pattern]")
    val dog = AnimalFactory.createAnimal("dog")
    val cat = AnimalFactory.createAnimal("cat")
    println("Dog: ${dog?.speak()}")
    println("Cat: ${cat?.speak()}")

    println()
    println("[Method Chain Pattern]")
    val calc = Calculator()

    val result1 = calc.add(5.0).getResult()
    println("add(5.0): $result1")

    val result2 = calc.multiply(2.0).getResult()
    println("multiply(2.0): $result2")

    val result3 = calc.subtract(3.0).getResult()
    println("subtract(3.0): $result3")

    val result4 = calc.divide(2.0).getResult()
    println("divide(2.0): $result4")

    val finalResult = calc.getResult()
    println("result: $finalResult")


    println()
    println("[Builder Pattern]")
    val person = PersonBuilder()
        .withName("홍길동")
        .withAge(20)
        .withAddress("부산대학교 양산캠퍼스")
        .build()
    println(person)

}

File: assignment_4/csharp/main.cs
using System;

class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("Adapter");
        ExternalClass external = new ExternalClass();
        IInternal adapter = new Adapter(external);
        Console.WriteLine(adapter.Fetch());


        Console.WriteLine("\nDecorator");
        Coffee coffee = new BasicCoffee();
        Console.WriteLine("coffee.cost() = " + coffee.Cost()); // 5

        Coffee coffee_with_milk = new MilkDecorator(coffee);
        Console.WriteLine("coffee_with_milk.cost() = " + coffee_with_milk.Cost()); // 7

        Coffee coffee_with_sugar = new SugarDecorator(coffee);
        Console.WriteLine("coffee_with_sugar.cost() = " + coffee_with_sugar.Cost()); // 6

        Coffee coffee_with_milk_and_sugar = new SugarDecorator(coffee_with_milk);
        Console.WriteLine("coffee_with_milk_and_sugar.cost() = " + coffee_with_milk_and_sugar.Cost()); // 8


        Console.WriteLine("\nFacade");
        Computer computer = new Computer();
        computer.Boot();


        Console.WriteLine("\nFactory Method");
        Animal dog = AnimalFactory.CreateAnimal("dog");
        Animal cat = AnimalFactory.CreateAnimal("cat");
        Console.WriteLine($"Dog says: {dog?.Speak()}");
        Console.WriteLine($"Cat says: {cat?.Speak()}");


        Console.WriteLine("\nMethod Chaining");
        Calculator calc = new Calculator();

        calc.Add(10.0);
        Console.WriteLine("add(10.0) = " + calc.GetResult());

        calc.Subtract(2.0);
        Console.WriteLine("subtract(2.0) = " + calc.GetResult());

        calc.Multiply(3.0);
        Console.WriteLine("multiply(3.0) = " + calc.GetResult());

        calc.Divide(4.0);
        Console.WriteLine("divide(4.0) = " + calc.GetResult());

        Console.WriteLine("getResult() = " + calc.GetResult());

        Console.WriteLine("Builder");
        var person = new PersonBuilder()
            .WithName("홍길동")
            .WithAge(20)
            .WithAddress("부산대학교 양산캠퍼스")
            .Build();
        Console.WriteLine(person);
        Console.WriteLine();


    }
}
